app/controllers/products_controller.rb

```diff
@@ -0,0 +1,93 @@
+class ProductsController < ApplicationController
+  before_action :find_product, only: [:show, :edit, :update, :retire]
+  before_action :require_login, except: [:index, :show]
+
+  def index
+    @products = Product.all
+  end
+
+  def show
+    if @product.nil?
+      redirect_to products_path
+      return
+    end
+  end
+
+  def new
+    @product = Product.new
+  end
+
+  def create
+    @product = Product.new(product_params)
+    @product.merchant = @current_merchant
+
+    if @product.save
+      redirect_to product_path(@product.id)
+      return
+    else
+      @product.errors.each do |type, err|
+        flash.now[type] = "Something went wrong: #{type}: #{err}"
+      end
+      render :new
+    end
+  end
+
+  def edit
+    if @product.nil?
+      redirect_to products_path
+      return
+    end
+  end
+
+  def update
+    if @product.nil?
+      flash[:error] = "Cannot find product to update"
+      redirect_to products_path
+      return
+    elsif @product.update(product_params)
+      redirect_to product_path(@product.id)
+      return
+    else
+      @product.errors.each do |type, err|
+        flash.now[type] = "Something went wrong: #{type}: #{err}"
+      end
+      render :edit
+      return
+    end
+  end
+
+
+  def retire
+
+    if @product.nil?
+      flash[:error] = "Cannot find product to retire"
+      redirect_to products_path
+      return
+    end
+    if @product.retired
+      @product.retired = false
+      @product.save
+      flash[:success] = "Put product back for sale!"
+      redirect_to product_path(@product)
+      return
+    else
+      @product.retired = true
+      @product.save
+      flash[:success] = "Retired product"
+      redirect_to product_path(@product)
+      return
+    end
+  end
+
+  private
+
+  def find_product
+    @product = Product.find_by(id: params[:id])
+  end
```

Comment:
The `flash` and redirect behavior you rolled one off above should probably have been moved into here:

```suggestion
  def find_product
    @product = Product.find_by(id: params[:id])
    
    if @product.nil?
      flash[:error] = "Cannot find product"
      redirect_to products_path
      return
    end
  end
```